src/main/java/org/littleshoot/proxy/impl/ServerGroup.java

```diff
@@ -219,6 +230,14 @@ private void shutdown(boolean graceful) {
             allEventLoopGroups.addAll(threadPools.getAllEventLoops());
         }
 
+        payloadProcessingExecutor.shutdown();
+
+        try {
+            payloadProcessingExecutor.awaitTermination(10, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            log.warn("Failed to shutdown payload processing executor properly", e);
```

Comment:
you must restore the interruption flag here. missing to do so will have have an adverse effect on the further shutdown code below 

```java

        try {
            payloadProcessingExecutor.awaitTermination(10, TimeUnit.SECONDS);
        } catch (InterruptedException e) {
            log.warn("Failed to shutdown payload processing executor properly", e);
        }

        for (EventLoopGroup group : allEventLoopGroups) {
            if (graceful) {
                group.shutdownGracefully();
            } else {
                group.shutdownGracefully(0, 0, TimeUnit.SECONDS);
            }
        }

        if (graceful) {
            for (EventLoopGroup group : allEventLoopGroups) {
                try {
                    group.awaitTermination(60, TimeUnit.SECONDS);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();

                    log.warn("Interrupted while shutting down event loop");
                }
            }
        }
```

Reply:
also, this shutdown sequence violates the recommended approach, as described in https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html

```java
 void shutdownAndAwaitTermination(ExecutorService pool) {
   pool.shutdown(); // Disable new tasks from being submitted
   try {
     // Wait a while for existing tasks to terminate
     if (!pool.awaitTermination(60, TimeUnit.SECONDS)) {
       pool.shutdownNow(); // Cancel currently executing tasks
       // Wait a while for tasks to respond to being cancelled
       if (!pool.awaitTermination(60, TimeUnit.SECONDS))
           System.err.println("Pool did not terminate");
     }
   } catch (InterruptedException ie) {
     // (Re-)Cancel if current thread also interrupted
     pool.shutdownNow();
     // Preserve interrupt status
     Thread.currentThread().interrupt();
   }
 }
```

Reply:
Fixed.